File: 2022/Python/Day 3/AoC-Day3-Part 2.py
def D3P2 (input):
  input_file = open(input,"r")
  input_list = input_file.read().splitlines()

  priority_dict = {}
  priority_num = 1
  for char in 'abcdefghijklmnopqrstuvwxyz':
    priority_dict[char] = priority_num
    priority_dict[char.upper()] = priority_num + 26
    priority_num += 1
  
  output = 0
  same_list = []
  for idx in range(int(len(input_list)/3)):
    bag1 = input_list[idx*3]
    bag2 = input_list[idx*3+1]
    bag3 = input_list[idx*3+2]
    for char in bag1:
      if char in bag2:
        if char in bag3:
          same_list.append(char)
          output += priority_dict[char]
          break
  

  return output

assert(D3P2('D3 Test.txt') == 70), "Function does not match example given"

if __name__ == '__main__':
  print(D3P2('D3 Input.txt'))

File: 2022/JavaScript/Day 3/AoC-Day3-Part2.js
// const { readFileSync } = require('fs');

// const D3P2 = (input) => {
//   const contents = readFileSync(input, 'utf-8');
//   const inputArr = contents.split(/\n/);

//   const priorityObj = {};
//   'abcdefghijklmnopqrstuvwxyz'.split('').forEach((ele, idx) => {
//     priorityObj[ele] = idx + 1;
//     priorityObj[ele.toUpperCase()] = idx + 27;
//   });

//   const groupArr = inputArr.reduce((acc, cur, idx) => {
//     if (idx % 3 === 0) {
//       acc.push([cur]);
//     } else {
//       acc[Math.floor(idx / 3)].push(cur);
//     }
//     return acc;
//   }, []);

//   const sameArr = groupArr.map((arr) =>
//     arr.reduce((acc, cur) => {
//       if (acc.length === 0) {
//         return acc.push(cur.split(''));
//       } else {
//         cur.split('').forEach((ele, idx) => {
//           if (acc.includes(ele)) {
//             acc.splice(acc.indexOf(ele));
//           }
//         });
//       }
//     }, [])
//   );

  // const sameArr = inputArr.map(line => {
  //   return line.split('').find( (char,idx,array) => array.slice(array.length/2).includes(char))
  // })

  // return groupArr;
  // sameArr.reduce((acc,cur) => priorityObj[cur]+acc,0)
// };

// const assert = (condition) => {
//   if (!condition) {
//     throw Error('Function does not match example given');
//   }
// };

// assert(D3P2('./D3 Test.txt')===70)
// console.log(D3P2('./D3 Test.txt'));
// console.log(D3P2('./D3 Input.txt'))
